feature_engine/feature_selection.py

```diff
@@ -88,3 +89,138 @@ def transform(self, X):
+        When drop_quasi_constant_features is True, features showing quasi_constant_threshold percentage of values
+        will be dropped
+
+    Attributes
+    ----------
+
+    constant_features_: list
+        List of constant features in a dataframe
+
+    quasi_constant_features_: list
+        List of quasi constant features in a dataframe
+
+    selected_features_: list
+        List of selected features which are non-constant and non-quasi constant
+
+    """
+    def __init__(self, drop_quasi_constant_features=False, quasi_constant_threshold=None):
+
+        self.drop_quasi_constant_features = drop_quasi_constant_features
+        self.quasi_constant_threshold = quasi_constant_threshold
+
+        self.constant_features_ = []
+        self.selected_features_ = []
+        self.quasi_constant_features_ = []
+        self._input_shape_ = ()
+
+        if not self.drop_quasi_constant_features and self.quasi_constant_threshold is not None:
+            warnings.warn(
+                'Setting a quasi_constant_threshold has no effect since drop_quasi_constant_features is False. You '
+                'should leave quasi_constant_threshold to its default (None), or set drop_quasi_constant_features=True.'
+                , FutureWarning
+            )
+
+        if self.drop_quasi_constant_features:
+            if self.quasi_constant_threshold <= 0 or self.quasi_constant_threshold >= 1:
+                raise ValueError("Value of threshold should be between 0 and 1")
+
+    def fit(self, X, y=None):
+
+        """
+        Find constant and non-quasi constant features from a dataframe
+
+        Parameters
+        ----------
+
+        X: pandas dataframe of shape = [n_samples, n_features]
+            The input dataframe
+
+        y: None
+            y is not needed for this transformer. You can pass y or None.
+
+        """
+
+        # data frame check
+        X = _is_dataframe(X)
+
+        # get input shape
+        self._input_shape_ = X.shape
+
+        # constant features
+        self.constant_features_ = [feat for feat in X.columns if X[feat].nunique() == 1]
+
+        # non-constant features
+        self.selected_features_ = [feat for feat in X.columns if feat not in self.constant_features_]
+
+        # if total constant features is equal to total features raise an error
+        if len(self.constant_features_) == len(X.columns):
+            raise ValueError("The resulting dataframe will have no columns after dropping all constant features")
+
+        if self.drop_quasi_constant_features:
+            # quasi-constant features
+            self.quasi_constant_features_ = [feat for feat in X.columns if X[feat].value_counts(
+                normalize=True, sort=True, ascending=False).iloc[0] > self.quasi_constant_threshold
+                                             and feat not in self.constant_features_
+                                             and feat in self.selected_features_]
```

Comment:
would it be possible to make this bit more readable? without so many ifs? I think the use of only the parameter threshold would help.

Reply:
If we use the following code wouldn't that capture constant feature in quasi_constant_feat list since predominant for the constant feature would be 1? We would ultimately need to remove constant features from the quasi_constant_feat list which does the first if statement. I would revisit ```and feat in self.selected_features_``` and see it's needed or not. 
Rest division by ```len``` and ```sort_values()``` is taken care by ```value_counts()``` arguments. What do you think? And I guess list comprehension would be faster than for loop
```
quasi_constant_feat = []
for feature in X_train.columns:

    # find the predominant value
    predominant = (X_train[feature].value_counts() / np.float(
        len(X_train))).sort_values(ascending=False).values[0]

    # evaluate predominant feature
    if predominant > 0.998:
        quasi_constant_feat.append(feature)

len(quasi_constant_feat)
````

Reply:
```
if tol==1:
    self.selected_features_ = [feat for feat in X.columns if feat not in self.constant_features_]

else: 
    self.selected_features_ = []
    for feature in X_train.columns:

        # find the predominant value
        predominant = (X_train[feature].value_counts() / np.float(
                 len(X_train))).sort_values(ascending=False).values[0]

       # evaluate predominant feature
       if predominant > self.tol:
           self.selected_features_.append(feature)
```
 That is the functionality I am thinking of. Would you agree. Only thing to check is if this would also capture constant or would through an error. But I think it should work

Reply:
Maybe there is a typo here
```
if predominant < self.tol:
    self.selected_features_.append(feature)
```

Please have a look at the final code here, ignore the indentation(copy+paste stuff) and if everything looks good I'll use this.
```


 class DropConstantAndQuasiConstantFeatures(TransformerMixin, BaseEstimator):

 """

 Drops constant and quasi constant variables from a dataframe.
 Constant variables are those variables which show only one value for all the observations.
 By default, all constant variables are dropped.
 Quasi-constant variables are those variables which are almost constant. In other words, those features which have
 the same value for a large subset of total observations.
 This transformer works for both numerical and categorical variables.

 Parameters
 ---------

 tol: float, default=1
 Features showing tol percentage of values will be dropped

 variables: list, default=None
 The list of variables to inspect. If None, the transformer will select all variables in the dataframe

 Attributes
 ----------

 selected_features_: list
 List of selected features which are non-constant and non-quasi constant

 """

 def __init__(self, tol=1, variables=None):

 if tol < 0 or tol > 1:
 raise ValueError("tol takes values between 0 and 1")

 self.tol = tol
 self.variables = _define_variables(variables)

 def fit(self, X, y=None):

 """
 Find constant and non-quasi constant features from a dataframe

 Parameters
 ----------

 X: pandas dataframe of shape = [n_samples, n_features]
 The input dataframe

 y: None
 y is not needed for this transformer. You can pass y or None.

 """

 # check input dataframe
 X = _is_dataframe(X)

 # find all variables or check that those entered are present in the dataframe
 self.variables = _find_all_variables(X, self.variables)

 if self.tol == 1:

 # constant features
 constant_features = [feat for feat in self.variables if X[feat].nunique() == 1]

 self.selected_features_ = [feat for feat in X.columns if feat not in constant_features]

 # if total constant features is equal to total features raise an error
 if len(constant_features) == len(X.columns):
 raise ValueError("The resulting dataframe will have no columns after dropping all constant features")

 else:
 constant_plus_quasi_constant_features = []
 self.selected_features_ = []

 for feat in self.variables:

 predominant = (X[feat].value_counts() / np.float(len(X))).sort_values(ascending=False).values[0]

 if predominant > self.tol:
 constant_plus_quasi_constant_features.append(feat)

 # if total constant features is equal to total features raise an error
 if len(constant_plus_quasi_constant_features) == len(X.columns):
 raise ValueError("The resulting dataframe will have no columns after dropping "
 "constant and quasi constant features")

 self.selected_features_ = [feat for feat in X.columns if feat not in constant_plus_quasi_constant_features]

 self.input_shape_ = X.shape

 return self

 def transform(self, X):
 """
 Drops the constant and non-quasi constant features from a dataframe and returns the dataframe with
 selected features.

 Parameters
 ----------
 X: pandas dataframe
 The input dataframe from which features will be dropped

 Returns
 -------
 X_transformed: pandas dataframe of shape = [n_samples, n_features - (constant_features+quasi constant features)]
 The transformed dataframe with the remaining subset of variables.

 """

 # check if fit is performed prior to transform
 check_is_fitted(self)

 # check if input is a dataframe
 X = _is_dataframe(X)

 # check if number of columns in test dataset matches to train dataset
 _check_input_matches_training_df(X, self.input_shape_[1])

 # returned selected features
 X = X[self.selected_features_].copy()

 return X
```

Reply:
It is easier to review code if you make it on the script and push them to the branch. You will also be able to know if it is working properly through the tests. Would you be able to make the changes in the code and push them to the branch? Then, we take it from there.

Reply:
Pushed, let me know for any changes. Thanks